### PR TITLE
Instance tracking logs and InstanceTracker Logger

### PR DESCRIPTION
**Pull request by pandashadopan**
## Changes proposed in this PR
- Added part of instance tracking logging 

## Why are we making these changes?
- To have logs for instance tracking

### CODE REVIEW

File: waiter/bin/run-using-minimesos.sh [hunk 1]
```diff
@@ -17,6 +17,8 @@ if [ ${EXIT_CODE} -eq 0 ]
 then
     export WAITER_MARATHON=${MINIMESOS_MARATHON%;}
     export WAITER_ZOOKEEPER_CONNECT_STRING=$(echo ${MINIMESOS_ZOOKEEPER} | awk -F/ '{print $3}')
+    export GRAPHITE_SERVER_PORT=5555
+    export JWKS_SERVER_URL=8040
```
Comment (DaoWen) [inline waiter/bin/run-using-minimesos.sh:21]: Should this just be the port number? Or should there be a `localhost` here?

File: waiter/src/waiter/scheduler.clj [hunk 2]
```diff
@@ -76,7 +76,8 @@
    ^String log-directory
    ^String message])
 
-(defn make-ServiceInstance [value-map]
+(defn make-ServiceInstance [{id :id started-at :started-at :as value-map}]
+  (log/log "InstanceTracker" :debug nil (str "Created an instance with id: " id " that started at " started-at))
```
Comment (DaoWen) [inline waiter/src/waiter/scheduler.clj:80]: We usually prefer creating a map with the fields we're interested in, and then logging the map. Same for extra data included in exception error messages.
```clojure
(log/log "InstanceTracker" :debug (str "Created an instance " {:id id :started-at started-at}))
```

File: waiter/src/waiter/handler.clj [hunk 3]
```diff
@@ -176,6 +176,10 @@
                                                            (when started-at
                                                              (du/str-to-date started-at)))))]
                     (notify-instance-killed-fn instance)))
+                (log/log "InstanceTracker" :debug nil (str "Instance " instance-id " BLACKLIST " (utils/clj->json
```
Comment (DaoWen) [inline waiter/src/waiter/handler.clj:179]: I would prefer if you declared a macro similar to `scheduler/log` that delegates to `(log/log "InstanceTracker" :debug nil ...)`.

File: waiter/src/waiter/scheduler.clj [hunk 4]
```diff
@@ -47,6 +47,11 @@
   [& args]
   `(log/log "Scheduler" :debug nil (print-str ~@args)))
 
+(defmacro logI
```
Comment (DaoWen) [inline waiter/src/waiter/scheduler.clj:50]: Let's rename this to `log-service-instance`.

File: waiter/src/waiter/scheduler.clj [hunk 5]
```diff
@@ -47,6 +47,11 @@
...
+  "Log InstanceTracker-specific messages."
+  [instance]
+  `(log/log "InstanceTracker" :debug nil (utils/clj->json instance)))
```
Comment (DaoWen) [inline waiter/src/waiter/scheduler.clj:53]: We want to `assoc` a couple of things here:
* `leader?`
* event-type (keepalive, running, killed, healthy, etc.)

File: waiter/src/waiter/scheduler/cook.clj [hunk 6]
```diff
@@ -308,6 +308,8 @@
                             allowed-users)
         failed-instances (map job->service-instance failed-jobs)]
     (when (seq failed-instances)
+      (doseq [failed-instance failed-instances]
+        (scheduler/logI (assoc failed-instance :event-type "FAIL")))
```
Comment (DaoWen) [inline waiter/src/waiter/scheduler/cook.clj:312]: `(scheduler/log-service-instance failed-instance "FAIL")`

File: waiter/src/waiter/scheduler.clj (hunk 5 above)
Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj]: We're not going to do the `leader?` field right now. Too much refactoring involved.

File: waiter/src/waiter/scheduler.clj [hunk 7]
```diff
@@ -889,10 +905,10 @@
                                   (concat healthy-instances))
             known-instance-ids' (->> known-instances' (map :id) set)
             previously-known-instance? (comp known-instance-ids :id)
-            new-instance-ids (->> known-instances'
-                                  (remove previously-known-instance?)
-                                  (sort instance-comparator)
-                                  (mapv :id))
+            new-instances (->> known-instances'
```
Comment (DaoWen) [inline waiter/src/waiter/scheduler.clj:908]: Let's just log all of the instances from this collection with `:launch` as the event-type.

File: waiter/src/waiter/state/maintainer.clj [hunk 8]
```diff
@@ -642,6 +642,13 @@
                                                        (if (seq new-instance-ids) (str "New healthy instances: " new-instance-ids ".") "")
```
Comment (DaoWen) [inline waiter/src/waiter/state/maintainer.clj:642]: We should be logging new healthy (and unhealthy) instances in our instance log.

File: waiter/src/waiter/state/maintainer.clj [hunk 9]
```diff
@@ -628,11 +628,20 @@
...
-                                                 prev-instance-ids (set (map :id (get service-id->healthy-instances service-id)))
-                                                 new-instance-ids (filterv (complement prev-instance-ids) curr-instance-ids)
-                                                 rem-instance-ids (filterv (complement curr-instance-ids) prev-instance-ids)
+                                           (let [curr-instances (set healthy-instances)
+                                                 prev-instances (set (get service-id->healthy-instances service-id))
+                                                 new-instances (filterv (complement prev-instances) curr-instances)
+                                                 rem-instances (filterv (complement curr-instances) prev-instances)
+                                                 new-instance-ids (mapv :id new-instances)
+                                                 rem-instance-ids (mapv :id rem-instances)
+
```
Comment (DaoWen) [inline waiter/src/waiter/state/maintainer.clj:637]: Let's remove this blank line.

```suggestion
```

File: waiter/src/waiter/state/maintainer.clj [hunk 10]
```diff
@@ -642,6 +651,13 @@
                                                        (if (seq new-instance-ids) (str "New healthy instances: " new-instance-ids ".") "")
                                                        (if (seq rem-instance-ids) (str "Removed healthy instances: " rem-instance-ids ".") "")
                                                        (if (seq unhealthy-instance-ids) (str "Unhealthy instances: " unhealthy-instance-ids ".") ""))))
+
```
Comment (DaoWen) [inline waiter/src/waiter/state/maintainer.clj:654]: Let's remove this blank line.

```suggestion
```

File: waiter/src/waiter/state/responder.clj [hunk 11]
```diff
@@ -470,10 +471,10 @@
...
       {:current-state' (if (= :blacklisted response-code)
                          (-> current-state
-                           ; mark instance as blacklisted and set the expiry time
-                           (update-in [:instance-id->state instance-id] sanitize-instance-state)
-                           (update-in [:instance-id->state instance-id] update-status-tag-fn #(conj % :blacklisted))
-                           (update-state-by-blacklisting-instance-fn cid instance-id blacklist-period-ms))
+                             ; mark instance as blacklisted and set the expiry time
+                             (update-in [:instance-id->state instance-id] sanitize-instance-state)
+                             (update-in [:instance-id->state instance-id] update-status-tag-fn #(conj % :blacklisted))
+                             (update-state-by-blacklisting-instance-fn cid instance-id blacklist-period-ms))
```
Comment (DaoWen) [inline waiter/src/waiter/state/responder.clj:477]: Shams and I have been debating what our standard indention style should be for threading macros. For now, I think's its best to not change it if you're not actually editing. Please un-indent these to revert of the whitespace change.

File: waiter/src/waiter/state/responder.clj [hunk 12]
```diff
@@ -482,9 +483,10 @@
   [{:keys [instance-id->state] :as current-state} update-instance-id->blacklist-expiry-time-fn update-status-tag-fn
    instance-id expiry-time]
   (log/info "unblacklisting instance" instance-id "as blacklist expired at" expiry-time)
-  (cond-> (update-instance-id->blacklist-expiry-time-fn current-state #(dissoc % instance-id))
```
Comment (DaoWen) [inline waiter/src/waiter/state/responder.clj:485]: `cond->` returns the first argument unmodified if all the conditions are falsy. The `when` version will return `nil` instead.

File: waiter/src/waiter/state/responder.clj [hunk 13]
```diff
@@ -470,10 +471,10 @@
...
       {:current-state' (if (= :blacklisted response-code)
                          (-> current-state
-                           ; mark instance as blacklisted and set the expiry time
-                           (update-in [:instance-id->state instance-id] sanitize-instance-state)
-                           (update-in [:instance-id->state instance-id] update-status-tag-fn #(conj % :blacklisted))
-                           (update-state-by-blacklisting-instance-fn cid instance-id blacklist-period-ms))
+                               ; mark instance as blacklisted and set the expiry time
+                               (update-in [:instance-id->state instance-id] sanitize-instance-state)
+                               (update-in [:instance-id->state instance-id] update-status-tag-fn #(conj % :blacklisted))
+                               (update-state-by-blacklisting-instance-fn cid instance-id blacklist-period-ms))
```
Comment (DaoWen) [inline waiter/src/waiter/state/responder.clj:477]: Whitespace here is off.

Comment (pandashadopan) [reply waiter/src/waiter/state/responder.clj]: Resolved

File: waiter/bin/run-using-k8s.sh [hunk 14]
```diff
@@ -8,6 +8,10 @@
 # Runs Waiter, configured to use the Kubernetes scheduler.
 
 export WAITER_PORT=${1:-9091}
+export GRAPHITE_SERVER_PORT=5555
+export JWKS_PORT=8040
+export JWKS_SERVER_URL="http://127.0.0.1:${JWKS_PORT}/keys"
+export WAITER_TEST_JWT_ACCESS_TOKEN_URL="http://127.0.0.1:${JWKS_PORT}/get-token?host={HOST}"
```
Comment (shamsimam) [inline waiter/bin/run-using-k8s.sh:14]: Please move the test script (for k8s and minimesos) changes to separate PRs.

File: waiter/bin/run-using-minimesos.sh [hunk 15]
```diff
@@ -17,6 +17,10 @@ if [ ${EXIT_CODE} -eq 0 ]
 then
     export WAITER_MARATHON=${MINIMESOS_MARATHON%;}
     export WAITER_ZOOKEEPER_CONNECT_STRING=$(echo ${MINIMESOS_ZOOKEEPER} | awk -F/ '{print $3}')
+    export GRAPHITE_SERVER_PORT=5555
+    export JWKS_PORT=8040
+    export JWKS_SERVER_URL="http://127.0.0.1:${JWKS_PORT}/keys"
+    export WAITER_TEST_JWT_ACCESS_TOKEN_URL="http://127.0.0.1:${JWKS_PORT}/get-token?host={HOST}"
```
Comment (shamsimam) [inline waiter/bin/run-using-minimesos.sh:23]: Please move the test script (for k8s and minimesos) changes to separate PRs.

File: waiter/src/waiter/scheduler.clj [hunk 16]
```diff
@@ -47,6 +47,11 @@
   [& args]
   `(log/log "Scheduler" :debug nil (print-str ~@args)))
 
+(defmacro log-service-instance
+  "Log InstanceTracker-specific messages."
+  [instance event-type]
+  `(log/log "InstanceTracker" :debug nil (utils/clj->json (assoc ~instance :timestamp (t/now) :event-type ~event-type))))
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler.clj:53]: Why is this set at `debug` by default?

File: waiter/src/waiter/scheduler/kubernetes.clj [hunk 17]
```diff
@@ -205,6 +205,7 @@
                                               ;; we don't include the exit code in failed instances that were killed by k8s.
                                               (not (killed-by-k8s? newest-failure))
                                               (assoc :exit-code (:exitCode newest-failure)))]
+                (scheduler/log-service-instance newest-failure-instance :fail)
                 (swap! service-id->failed-instances-transient-store
                        update-in [service-id] assoc newest-failure-id newest-failure-instance)))))))
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/kubernetes.clj:210]: We should wrap the `service-id->failed-instances-transient-store` update in a function and that function can log the service instance. E.g. that function can be reused in the marathon scheduler below.

File: waiter/src/waiter/scheduler/shell.clj [hunk 18]
```diff
@@ -363,12 +363,15 @@
...
       (release-port! port->reservation-atom port port-grace-period-ms)
-      (assoc instance
-        :exit-code exit-value
-        :failed? (if (zero? exit-value) false true)
-        :healthy? false
-        :killed? true ; does not actually mean killed -- using this to mark inactive
-        :message message))
+      (let [newly-constructed-instance (assoc instance
+                                         :exit-code exit-value
+                                         :failed? (if (zero? exit-value) false true)
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/shell.clj:368]: Extract `failed?` into a local variable?

File: waiter/src/waiter/state/maintainer.clj [hunk 19]
```diff
@@ -628,11 +628,19 @@
                                                                               (dissoc service-id->instability-issue service-id))]
                                          (when (or (not= (get service-id->healthy-instances service-id) healthy-instances)
                                                    (not= (get service-id->unhealthy-instances service-id) unhealthy-instances))
-                                           (let [curr-instance-ids (set (map :id healthy-instances))
-                                                 prev-instance-ids (set (map :id (get service-id->healthy-instances service-id)))
-                                                 new-instance-ids (filterv (complement prev-instance-ids) curr-instance-ids)
-                                                 rem-instance-ids (filterv (complement curr-instance-ids) prev-instance-ids)
+                                           (let [curr-instances (set healthy-instances)
+                                                 prev-instances (set (get service-id->healthy-instances service-id))
+                                                 new-instances (filterv (complement prev-instances) curr-instances)
```
Comment (shamsimam) [inline waiter/src/waiter/state/maintainer.clj:633]: These can remain as `filter` instead of `filterv`

File: waiter/src/waiter/state/maintainer.clj [hunk 20]
```diff
@@ -628,11 +628,19 @@
...
+                                                 rem-instances (filterv (complement curr-instances) prev-instances)
```
Comment (shamsimam) [inline waiter/src/waiter/state/maintainer.clj:634]: These can remain as `filter` instead of `filterv`

File: waiter/src/waiter/state/maintainer.clj [hunk 21]
```diff
@@ -628,11 +628,19 @@
...
+                                                 new-instance-ids (mapv :id new-instances)
+                                                 rem-instance-ids (mapv :id rem-instances)
+                                                 prev-unhealthy-instances (set (get service-id->unhealthy-instances service-id))
+                                                 new-unhealthy-instances (filterv (complement prev-unhealthy-instances) unhealthy-instances)
```
Comment (shamsimam) [inline waiter/src/waiter/state/maintainer.clj:638]: These can remain as `filter` instead of `filterv`

File: waiter/src/waiter/state/maintainer.clj [hunk 22]
```diff
@@ -711,7 +725,9 @@
       {:go-chan go-chan
        :notify-instance-killed-fn (fn notify-router-state-maintainer-of-instance-killed [instance]
                                     (log/info "received notification of killed instance" (:id instance))
-                                    (async/go (async/>! kill-notification-chan {:instance instance})))
+                                    (scheduler/log-service-instance instance :kill)
+                                    (async/go
+                                      (async/>! kill-notification-chan {:instance instance})))
```
Comment (shamsimam) [inline waiter/src/waiter/state/maintainer.clj:730]: Nitpick: unnecessary formatting

File: waiter/src/waiter/state/maintainer.clj [hunk 23]
```diff
@@ -711,7 +725,9 @@
       {:go-chan go-chan
        :notify-instance-killed-fn (fn notify-router-state-maintainer-of-instance-killed [instance]
                                     (log/info "received notification of killed instance" (:id instance))
-                                    (async/go (async/>! kill-notification-chan {:instance instance})))
+                                    (scheduler/log-service-instance instance :kill)
```
Comment (shamsimam) [inline waiter/src/waiter/state/maintainer.clj:728]: The event is `kill-notification`. The `kill` has been triggered elsewhere, it should be added in the scheduler implementation.

File: waiter/src/waiter/state/responder.clj [hunk 24]
```diff
@@ -482,9 +483,14 @@
   [{:keys [instance-id->state] :as current-state} update-instance-id->blacklist-expiry-time-fn update-status-tag-fn
    instance-id expiry-time]
   (log/info "unblacklisting instance" instance-id "as blacklist expired at" expiry-time)
-  (cond-> (update-instance-id->blacklist-expiry-time-fn current-state #(dissoc % instance-id))
-    (contains? instance-id->state instance-id)
-    (update-in [:instance-id->state instance-id] update-status-tag-fn #(disj % :blacklisted))))
```
Comment (shamsimam) [inline waiter/src/waiter/state/responder.clj:487]: I'd prefer this same structure. We can wrap the `update-in` into a helper function that adds the necessary instance tracking logging. The `if-else` feels clunky here.

File: waiter/src/waiter/state/responder.clj [hunk 25]
```diff
@@ -482,9 +483,14 @@
...
+  (let [update-instance-id->blacklist-expiry-time-fn-ret
+        (update-instance-id->blacklist-expiry-time-fn current-state #(dissoc % instance-id))]
+    (if (contains? instance-id->state instance-id)
+      (do
+        (scheduler/log-service-instance {:id instance-id} :readmit)
```
Comment (shamsimam) [inline waiter/src/waiter/state/responder.clj:490]: Did we settle on the `readmit` and `eject` terms?

File: waiter/src/waiter/state/maintainer.clj [hunk 26]
```diff
@@ -628,11 +628,19 @@
...
-                                           (let [curr-instance-ids (set (map :id healthy-instances))
-                                                 prev-instance-ids (set (map :id (get service-id->healthy-instances service-id)))
-                                                 new-instance-ids (filterv (complement prev-instance-ids) curr-instance-ids)
-                                                 rem-instance-ids (filterv (complement curr-instance-ids) prev-instance-ids)
+                                           (let [curr-instances (set healthy-instances)
+                                                 prev-instances (set (get service-id->healthy-instances service-id))
+                                                 new-instances (filterv (complement prev-instances) curr-instances)
+                                                 rem-instances (filterv (complement curr-instances) prev-instances)
+                                                 new-instance-ids (mapv :id new-instances)
+                                                 rem-instance-ids (mapv :id rem-instances)
```
Comment (shamsimam) [inline waiter/src/waiter/state/maintainer.clj:636]: We should track instances that got removed.

File: waiter/src/waiter/scheduler.clj (hunk 16 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj:53]: To preserve the convention that Scheduler specific messages use debug by default. 

```
(defmacro log
  "Log Scheduler-specific messages."
  [& args]
  `(log/log "Scheduler" :debug nil (print-str ~@args)))
```

File: waiter/src/waiter/state/responder.clj (hunk 25 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/responder.clj:490]: No. We can change them later

Comment (DaoWen) [reply waiter/src/waiter/state/responder.clj:490]: We definitely had consensus on "eject", but we didn't really discuss the symmetric term.

File: waiter/src/waiter/state/maintainer.clj [hunk 27]
```diff
@@ -642,6 +650,12 @@
                                                        (if (seq new-instance-ids) (str "New healthy instances: " new-instance-ids ".") "")
                                                        (if (seq rem-instance-ids) (str "Removed healthy instances: " rem-instance-ids ".") "")
                                                        (if (seq unhealthy-instance-ids) (str "Unhealthy instances: " unhealthy-instance-ids ".") ""))))
+                                         (when (not= (get service-id->expired-instances service-id) expired-instances)
```
Comment (DaoWen) [inline waiter/src/waiter/state/maintainer.clj:653]: Since this logic is only used for InstanceTracker logging, let's predicate this extra work on the logger actually being enabled.

https://clojure.github.io/tools.logging/#clojure.tools.logging/enabled?

File: waiter/src/waiter/state/maintainer.clj (hunk 23 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/maintainer.clj]: Done

File: waiter/src/waiter/state/maintainer.clj (hunk 22 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/maintainer.clj]: Changed

File: waiter/src/waiter/state/maintainer.clj (hunk 21 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/maintainer.clj]: Done

File: waiter/src/waiter/state/maintainer.clj (hunk 20 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/maintainer.clj]: Done

File: waiter/src/waiter/state/maintainer.clj (hunk 19 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/maintainer.clj]: Done

File: waiter/src/waiter/scheduler/shell.clj (hunk 18 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler/shell.clj]: Done

File: waiter/src/waiter/scheduler/kubernetes.clj (hunk 17 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler/kubernetes.clj]: Done

File: waiter/bin/run-using-minimesos.sh (hunk 15 above)
Comment (pandashadopan) [reply waiter/bin/run-using-minimesos.sh]: Done

File: waiter/bin/run-using-k8s.sh (hunk 14 above)
Comment (pandashadopan) [reply waiter/bin/run-using-k8s.sh]: Done

File: waiter/src/waiter/state/responder.clj (hunk 24 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/responder.clj:487]: Done

File: waiter/src/waiter/state/maintainer.clj (hunk 27 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/maintainer.clj]: Done

File: waiter/src/waiter/scheduler.clj [hunk 28]
```diff
@@ -47,6 +47,16 @@
...
+
+(defn swap-atom-with-instance-logging
```
Comment (DaoWen) [inline waiter/src/waiter/scheduler.clj:55]: This function name should end with `!` (per Clojure conventions) since it mutates an atom.

Comment (shamsimam) [reply waiter/src/waiter/scheduler.clj:55]: Also, we should be specific about the fields we want to render rather than rendering the entire instance map. E.g. the request log does this where it extracts specific fields from the request/response.

Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj:55]: We should delegate to `add-instance-to-buffered-collection!` here. I think us *not* using that function in kubernetes.clj was a bug that I introduced.

Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj:55]: I would prefer our helper function have a simpler name, like `track-failed-instance!`

File: waiter/src/waiter/scheduler.clj [hunk 29]
```diff
@@ -754,7 +754,9 @@
 (defn add-instance-to-transient-store!
   [transient-store service-id instance event-type & args]
   (log-service-instance instance event-type)
-  (add-instance-to-buffered-collection! transient-store service-id instance args))
+  (if (= (count args) 3)
+    (add-instance-to-buffered-collection! transient-store service-id instance (nth args 0) (nth args 1) (nth args 2))
+    (add-instance-to-buffered-collection! transient-store service-id instance (nth args 0))))
```
Comment (pandashadopan) [inline waiter/src/waiter/scheduler.clj:759]: Should I use (first ) or (nth ... 0) in this case? I used nth to preserve the fact that nth is used on the line above.

Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj:759]:
```suggestion
    (apply add-instance-to-buffered-collection! transient-store service-id instance args)))
```

File: waiter/src/waiter/scheduler/kubernetes.clj [hunk 30]
```diff
@@ -205,8 +205,8 @@
                                               ;; we don't include the exit code in failed instances that were killed by k8s.
                                               (not (killed-by-k8s? newest-failure))
                                               (assoc :exit-code (:exitCode newest-failure)))]
-                (swap! service-id->failed-instances-transient-store
-                       update-in [service-id] assoc newest-failure-id newest-failure-instance)))))))
+                (scheduler/add-instance-to-transient-store! service-id->failed-instances-transient-store
+                                             service-id newest-failure-instance :fail newest-failure-id)))))))
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler/kubernetes.clj:209]: Nitpick: Indentation is off here.

Comment (pandashadopan) [reply waiter/src/waiter/scheduler/kubernetes.clj]: Done

File: waiter/src/waiter/scheduler.clj (hunk 16 above)
Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj:53]: OK, I like the idea of setting all of our current messages to `:info` level, and reserving the ability to set some more noisy events later to `:debug` level. @pandashadopan , let's met the log level an explicit argument to this macro.

I also with Shams that we should `select-keys` on the instance to choose the fields we want to log.

File: waiter/src/waiter/scheduler.clj [hunk 31]
```diff
@@ -730,13 +735,25 @@
 
 (defn add-instance-to-buffered-collection!
   "Helper function to add/remove entries into the transient store"
-  [transient-store max-instances-to-keep service-id instance-entry initial-value-fn remove-fn]
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler.clj:733]: Why did we reorder the argument? It has triggered a bunch of changes in the callers.

File: waiter/src/waiter/scheduler.clj [hunk 32]
```diff
@@ -730,13 +735,25 @@
...
+  ([transient-store service-id instance-entry max-instances-to-keep initial-value-fn remove-fn]
+   (swap! transient-store
+          (fn [service-id->failed-instances]
+            (update-in service-id->failed-instances [service-id]
+                       #(cond-> (or % (initial-value-fn))
+                                (= max-instances-to-keep (count %)) (remove-fn)
+                                true (conj instance-entry))))))
+  ;; Since we have a map (failure-id to failure-instance) in k8s we just update
+  ;; the map instead of adding element to a set like in marathon's case
+  ([transient-store service-id instance-entry failure-id]
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler.clj:747]: I think we should not have this override and force the caller to provide a limit.
Also, the caller can assoc the `failure-id` into the instance.

File: waiter/src/waiter/scheduler.clj (hunk 31 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj:733]: I reordered the argument because I wanted to use & args in the `add-instance-to-transient-store!` without argument reordering it would be less idiomatic:

```
(defn add-instance-to-transient-store!
  [transient-store service-id instance event-type & args]
  (log-service-instance instance event-type)
  (apply add-instance-to-buffered-collection! transient-store service-id instance args))
```

Comment (shamsimam) [reply waiter/src/waiter/scheduler.clj:733]: If we don't have the override we don't have a need to use `args`.

Comment (shamsimam) [reply waiter/src/waiter/scheduler.clj:733]: Please do not change the signature of this function, `max-instances-to-keep` should be passed by the caller.

File: waiter/src/waiter/scheduler.clj [hunk 33]
```diff
@@ -730,13 +735,25 @@
...
+  ([transient-store service-id instance-entry]
+   (swap! transient-store
+          (fn [service-id->failed-instances]
+            (update-in service-id->failed-instances [service-id]
+                       #(assoc % (:failure-id instance-entry) instance-entry))))))
+
+(defn add-instance-to-transient-store!
+  [transient-store service-id instance event-type & args]
+  (log-service-instance instance event-type)
+  (apply add-instance-to-buffered-collection! transient-store service-id instance args))
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler.clj:756]: We should not need to use `& args` and `apply` in this function.

File: waiter/src/waiter/scheduler.clj (hunk 31 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj:732]: Resolved

Comment (shamsimam) [reply waiter/src/waiter/scheduler.clj:732]: I still see a diff. There is no need for the extra set of prenthesis since we have only one signature for the function now.

File: waiter/src/waiter/state/responder.clj [hunk 34]
```diff
@@ -30,7 +30,8 @@
             [waiter.util.date-utils :as du]
             [waiter.util.utils :as utils]
             [waiter.work-stealing :as work-stealing])
-  (:import clojure.lang.PersistentQueue))
+  (:import clojure.lang.PersistentQueue
+           (org.joda.time DateTime)))
```
Comment (shamsimam) [inline waiter/src/waiter/state/responder.clj:34]: Is this import unused? If so, please remove.

File: waiter/src/waiter/state/responder.clj [hunk 35]
```diff
@@ -478,13 +479,21 @@
        :response-chan response-chan
        :response response-code})))
 
+(defn- update-in-with-log-instance
+  [mp log-body event-type log-level key-sequence func & args]
```
Comment (shamsimam) [inline waiter/src/waiter/state/responder.clj:483]: Based on the single invocation below isn't `log-body` always the `instance-id`?
We should avoid the use of `apply` whenever we can.

Comment (pandashadopan) [reply waiter/src/waiter/state/responder.clj:483]: I added the abstract layer so that we can reuse this function later. Should I make it only consider the current case and we would change it later if we would need to?

Comment (shamsimam) [reply waiter/src/waiter/state/responder.clj:483]: Yes, please change it to consider the current case.

File: waiter/src/waiter/scheduler.clj [hunk 36]
```diff
@@ -47,6 +47,11 @@
   [& args]
   `(log/log "Scheduler" :debug nil (print-str ~@args)))
 
+(defmacro log-service-instance
+  "Log InstanceTracker-specific messages."
+  [instance event-type log-level]
+  `(log/log "InstanceTracker" ~log-level nil (utils/clj->json (assoc ~instance :timestamp (t/now) :event-type ~event-type))))
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler.clj:53]: Reminder: https://github.com/twosigma/waiter/pull/1113#discussion_r442259899

File: waiter/src/waiter/state/responder.clj (hunk 34 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/responder.clj]: Removed

File: waiter/src/waiter/state/responder.clj (hunk 35 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/responder.clj]: Done

File: waiter/src/waiter/scheduler.clj (hunk 36 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done

File: waiter/src/waiter/scheduler.clj [hunk 37]
```diff
@@ -47,6 +47,11 @@
   [& args]
   `(log/log "Scheduler" :debug nil (print-str ~@args)))
 
+(defmacro log-service-instance
+  "Log InstanceTracker-specific messages."
+  [instance event-type log-level key-seq]
+  `(log/log "InstanceTracker" ~log-level nil (utils/clj->json (assoc (select-keys ~instance ~key-seq) :timestamp (t/now) :event-type ~event-type))))
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler.clj:53]: The `key-seq` should be determined in this function. The caller should not be the one passing this list in.

File: waiter/src/waiter/state/responder.clj [hunk 38]
```diff
@@ -478,13 +478,18 @@
        :response-chan response-chan
        :response response-code})))
 
+(defn- update-in-with-log-instance
+  [mp log-body event-type log-level key-sequence update-fn func]
```
Comment (shamsimam) [inline waiter/src/waiter/state/responder.clj:482]: `mp` is better referred to as the `current-state`.
`key-sequence` is actually  better referred to as the `path`.

Importantly, if this function is named `clear-rejection-state` (as that is the only use currently), then none of the arguments except current state (`mp`) and `instance-id` need to be passed to this function.

File: waiter/src/waiter/scheduler.clj (hunk 37 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj:53]:
```
(let key-seq [:id :service-id :started-at :healthy? :health-check-status :flags :exit-code :host :port :extra-ports :log-directory :message]
  (log/log ...))
```

Would something like this suffice?

Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj:53]: We should just `def` the key-seq outside the macro and reference it inside.

Also, `:flags` needs to be flattened, otherwise it won't show correctly up in ELK.

Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj:53]: `:extra-ports` is also an array that needs to be flattened.

Comment (shamsimam) [reply waiter/src/waiter/scheduler.clj:53]: > How would we request certain keys from the caller side in this case?

The caller shouldn't have to care. This function will extract whatever keys it feels are necessary to log the instance and event.

Comment (DaoWen) [reply waiter/src/waiter/scheduler.clj:53]: I was thinking about `:extra-ports` again, and I think we should just log out the count. Since they are always sequential in both the Marathon and Kubernetes implementations, having port0 and the extra-ports count is sufficient.

File: waiter/src/waiter/scheduler.clj [hunk 39]
```diff
@@ -47,6 +47,11 @@
   [& args]
   `(log/log "Scheduler" :debug nil (print-str ~@args)))
 
+(defmacro log-service-instance
```
Comment (shamsimam) [inline waiter/src/waiter/scheduler.clj:50]: Can we have this as a function instead of a macro?

I do not expect instance logging events to be that frequent. In any case, they are not on the reserve instance hot paths.

Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done

File: waiter/src/waiter/state/responder.clj (hunk 38 above)
Comment (pandashadopan) [reply waiter/src/waiter/state/responder.clj]: Done

File: waiter/src/waiter/scheduler.clj (hunk 37 above)
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done

File: waiter/src/waiter/scheduler.clj [hunk 40]
```diff
@@ -918,6 +955,9 @@
             (when leader?
               (statsd/histo! metric-group "startup_time"
                              (du/interval->nanoseconds duration)))))
+        (doseq [new-instance new-instances]
+          (when (healthy-instances? new-instance)
```
[the remark this answers is not in the document]
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done

File: waiter/src/waiter/scheduler.clj [hunk 41]
```diff
@@ -737,6 +768,11 @@
                          (= max-instances-to-keep (count %)) (remove-fn)
                          true (conj instance-entry))))))
 
+(defn add-to-store-and-track-instance!
+  [transient-store max-instances-to-keep service-id instance event-type log-level initial-value-fn remove-fn]
```
[the remark this answers is not in the document]
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done

File: waiter/src/waiter/scheduler.clj [hunk 42]
```diff
@@ -47,6 +47,37 @@
...
+                       :flags
+                       (map (fn [flag] [flag true]))
+                       (into {}))]
+    (log/log "InstanceTracker" log-level nil (utils/clj->json
+                                               (assoc (merge
+                                                        (dissoc instance-to-log :flags)
+                                                        flags-map)
+                                                 :extra-ports extra-ports
+                                                 :timestamp (t/now)
+                                                 :event-type event-type)))))
```
[the remark this answers is not in the document]
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done

File: waiter/src/waiter/scheduler.clj [hunk 43]
```diff
@@ -47,6 +47,37 @@
...
+              :blackist-period-ms])
+
+(defn log-service-instance
+  "Log InstanceTracker-specific messages."
+  [instance event-type log-level]
+  (let [extra-ports (count (:extra-ports instance))
+        instance-to-log (select-keys instance key-seq)
+        flags-map (->> instance-to-log
+                       :flags
+                       (map (fn [flag] [flag true]))
```
[the remark this answers is not in the document]
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done

File: waiter/src/waiter/scheduler.clj [hunk 44]
```diff
@@ -47,6 +47,37 @@
   [& args]
   `(log/log "Scheduler" :debug nil (print-str ~@args)))
 
+(def key-seq [:id
```
[the remark this answers is not in the document]
Comment (pandashadopan) [reply waiter/src/waiter/scheduler.clj]: Done
